drivers/spi/ice40.c

```diff
@@ -0,0 +1,411 @@
+ * Name: ice40_configspi
+ *
+ * Description:
+ *   Configure the SPI instance for to match the DAT-31R5-SP+
+ *   specifications
+ *
+ ****************************************************************************/
+
+static inline void
+ice40_configspi (FAR struct spi_dev_s *spi)
+{
+  DEBUGASSERT (spi != NULL);
+
+  /* Configure SPI Mode for the ICE40 */
+
+  SPI_SETMODE (spi, ICE40_SPI_MODE);
+  SPI_SETBITS (spi, 8);
+
+  SPI_HWFEATURES (spi, 0);
+  SPI_SETFREQUENCY (spi, CONFIG_ICE40_SPI_FREQUENCY);
+}
+
+/****************************************************************************
+ * Name: ice40_init_fpga
+ *
+ * Description:
+ *  Initialize the FPGA - set it to SPI Master load mode
+ *  Reset the FPGA with the CS pin active low
+ *  and send 8 dummy bits with CS high to start the SPI transfer.
+ *
+ ****************************************************************************/
+
+static int
+ice40_init_fpga (FAR struct ice40_dev_s *dev)
+{
+  DEBUGASSERT (dev != NULL);
+  DEBUGASSERT (dev->spi != NULL);
+
+  SPI_LOCK (dev->spi, true);
+
+  ice40_configspi (dev->spi);
+
+  dev->ops->reset (dev, true);
+  up_udelay (2);
+  dev->ops->select (dev, true);
+  up_udelay (2);
+  dev->ops->reset (dev, false);
+  up_udelay (1200);
+
+  dev->ops->select (dev, false);
+  SPI_SEND (dev->spi, 0xff);
+  dev->ops->select (dev, true);
+
+  dev->in_progress = true;
+
+  return 0;
+}
+
+/****************************************************************************
+ * Name: ice_v_writeblk
+ *
+ * Description:
+ *  Write block to the ICE40 FPGA, max 4096 bytes
+ ****************************************************************************/
+
+static inline int
+ice40_writeblk (FAR struct ice40_dev_s *dev, FAR const char *buffer,
+                size_t buflen)
+{
+  uint32_t nbytes;
+
+  DEBUGASSERT (dev != NULL);
+  DEBUGASSERT (dev->spi != NULL);
+
+  DEBUGASSERT (buffer != NULL);
+  DEBUGASSERT (buflen > 0);
+
+  if (!dev->in_progress)
+    {
+      _err ("ERROR: FPGA not initialized\n");
+      return -EINVAL;
+    }
+
+  ice40_configspi (dev->spi);
+
+  while (buflen > 0)
+    {
+      nbytes = buflen;
+      if (nbytes >= ICE_SPI_MAX_XFER)
+        {
+          nbytes = ICE_SPI_MAX_XFER;
+        }
+
+      SPI_SNDBLOCK (dev->spi, buffer, nbytes);
+
+      buffer += nbytes;
+      buflen -= nbytes;
+    }
+
+  return 0;
+}
+
+/****************************************************************************
+ * Name: ice_v_endwrite
+ *
+ * Description:
+ *  End writing bitstream to the ICE40 FPGA
+ ****************************************************************************/
+
+static int
+ice40_endwrite (FAR struct ice40_dev_s *dev)
+{
+  ice40_configspi (dev->spi);
+  int cdone = 0;
+
+  DEBUGASSERT (dev != NULL);
+  DEBUGASSERT (dev->spi != NULL);
+
+  if (!dev->in_progress)
+    {
+      _err ("ERROR: FPGA not initialized\n");
+      return -EINVAL;
+    }
+
+  dev->ops->select (dev, false);
+
+  for (size_t i = 0; i < 20; i++)
```

Comment:
Instead using a hard-coded value 20, please create a macro to indicate the value meaning.